### PR TITLE
add datahub DataSourceV2

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

### CODE REVIEW

File: assembly/pom.xml [hunk 1]
```diff
@@ -109,6 +116,11 @@
                             <include>com.aliyun.emr:emr-logservice_${scala.binary.version}</include>
                             <include>com.aliyun.emr:emr-jdbc_${scala.binary.version}</include>
                             <include>com.aliyun.emr:emr-hbase_${scala.binary.version}</include>
+                            <include>com.aliyun.emr:emr-datahub_${scala.binary.version}</include>
+                            <include>com.101tec:zkclient</include>
```
Comment (wenxuanguan) [inline assembly/pom.xml]: why this lib is needed, since no code depend on zookeeper client

**Comment by wenxuanguan** [comment]
please add datahub example

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubOffsetReader.scala [hunk 2]
```diff
@@ -115,22 +116,54 @@ class DatahubOffsetReader(readerOptions: Map[String, String]) extends Logging {
...
         val cursor = datahubClient.getCursor(project, topic, datahubShard.shardId, CursorType.LATEST)
         (datahubShard, cursor.getSequence)
-      }}.toMap
+      }.toMap
     }
   }
 
+  def fetchLatestOffsets(knownOffsets: Option[Map[DatahubShard, Long]]): Map[DatahubShard, Long] =
+    runUninterruptibly {
+      withRetriesWithoutInterrupt {
```
Comment (wenxuanguan) [inline emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubOffsetReader.scala]: some duplicate code with fetchLatestOffsets

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubRecordToUnsafeRowConverter.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  private val scale = sourceOptions("decimal.scale").toInt
+
+  def toUnsafeRow(
+      record: RecordEntry,
+      project: String,
+      topic: String,
+      shardId: String): UnsafeRow = {
+    rowWriter.reset()
+    rowWriter.zeroOutNullBytes()
+    if (!fallback) {
```
Comment (wenxuanguan) [inline emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubRecordToUnsafeRowConverter.scala]: really need fallback?

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubOffsetReader.scala (hunk 2 above)
Comment (uncleGen) [reply emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubOffsetReader.scala]: ok

**Pull request merged**